src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs

```diff
@@ -669,7 +687,7 @@ Exception GetInvalidArgumentsException()
             }
         }
 
-        private void ThrowInvalidArguments(int offset, int length)
+        private void ThrowInvalidArguments(int offset, int length, bool offsetIsStart = false)
```

Comment:
Can this instead take the `ExceptionArgument` to use? I find this naming pretty confusing.

You also didn't modify:

```
return ThrowHelper.GetArgumentOutOfRangeException(ExceptionArgument.offset);
```

So if `offsetIsStart == true`, but `hasValue == false`, this will throw for the argument `offset`.